[Feature request] Support for naming of StepVerifier scenario

Raising issue as proposed by @simonbasle in chat

Consider scenario 
```java
StepVerifier.create(Flux.never())
                  .expectError()
                  .verify(Duration.ofSeconds(10));
```

### Expected behavior
It would be nice to have descriptive message for failed scenario as a whole.
Proposed change is handy when multiple scenarios are verified 
in single test method, because It is not clear which one has failed.
E.g.

> java.lang.AssertionError: **[Scenario Request-check]** VerifySubscriber timed   
out on reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain@1a38c59b

### Actual behavior

>java.lang.AssertionError: VerifySubscriber timed out on
reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain@1a38c59b

### Steps to reproduce

### Reactor Core version
3.1.3.RELEASE

### JVM version (e.g. `java -version`)

java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)

--- a/reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java
+++ b/reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java
@@ -77,6 +77,12 @@ final class DefaultStepVerifierBuilder<T>
 	 */
 	static Duration defaultVerifyTimeout = StepVerifier.DEFAULT_VERIFY_TIMEOUT;
 
+	/**
+	 * The {@link ErrorFormatter} used for cases where no scenario name has been provided
+	 * through {@link StepVerifierOptions}.
+	 */
+	static final ErrorFormatter NO_NAME_ERROR_FORMATTER = new ErrorFormatter(null);
+
 	static void checkPositive(long n) {
 		if (n < 0) {
 			throw new IllegalArgumentException("'n' should be >= 0 but was " + n);
@@ -113,7 +119,7 @@ final class DefaultStepVerifierBuilder<T>
 			@Nullable Supplier<? extends Publisher<? extends T>> sourceSupplier) {
 		this.initialRequest = options.getInitialRequest();
 		this.options = options;
-		this.errorFormatter = new ErrorFormatter(options.getScenarioName());
+		this.errorFormatter = options.getScenarioName() == null ? NO_NAME_ERROR_FORMATTER : new ErrorFormatter(options.getScenarioName());
 		this.vtsLookup = options.getVirtualTimeSchedulerSupplier();
 		this.sourceSupplier = sourceSupplier;
 		this.script = new ArrayList<>();

--- a/reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java
+++ b/reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java
@@ -97,16 +97,13 @@ final class DefaultStepVerifierBuilder<T>
 		return new DefaultStepVerifierBuilder<>(options, scenarioSupplier);
 	}
 
-	@SuppressWarnings("unchecked")
-	static <T> SignalEvent<T> defaultFirstStep() {
-		return (SignalEvent<T>) DEFAULT_ONSUBSCRIBE_STEP;
-	}
-
-	final List<Event<T>>                             script;
-	final long                                       initialRequest;
-	final Supplier<? extends VirtualTimeScheduler>   vtsLookup;
-	final Supplier<? extends Publisher<? extends T>> sourceSupplier;
-	private final StepVerifierOptions options;
+	final         SignalEvent<T>                             defaultFirstStep;
+	final         List<Event<T>>                             script;
+	final         ErrorFormatter                             errorFormatter;
+	final         long                                       initialRequest;
+	final         Supplier<? extends VirtualTimeScheduler>   vtsLookup;
+	final         Supplier<? extends Publisher<? extends T>> sourceSupplier;
+	private final StepVerifierOptions                        options;
 
 	long hangCheckRequested;
 	int  requestedFusionMode = -1;
@@ -116,10 +113,12 @@ final class DefaultStepVerifierBuilder<T>
 			@Nullable Supplier<? extends Publisher<? extends T>> sourceSupplier) {
 		this.initialRequest = options.getInitialRequest();
 		this.options = options;
+		this.errorFormatter = new ErrorFormatter(options.getScenarioName());
 		this.vtsLookup = options.getVirtualTimeSchedulerSupplier();
 		this.sourceSupplier = sourceSupplier;
 		this.script = new ArrayList<>();
-		this.script.add(defaultFirstStep());
+		this.defaultFirstStep = newOnSubscribeStep(errorFormatter, "defaultOnSubscribe");
+		this.script.add(defaultFirstStep);
 
 		this.hangCheckRequested = initialRequest;
 	}
@@ -140,7 +139,7 @@ final class DefaultStepVerifierBuilder<T>
 		Objects.requireNonNull(assertionConsumer, "assertionConsumer");
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnError()) {
-				return fail(se, "expected: onError(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onError(); actual: %s", signal);
 			}
 			else {
 				try {
@@ -148,7 +147,7 @@ final class DefaultStepVerifierBuilder<T>
 					return Optional.empty();
 				}
 				catch (AssertionError e) {
-					if (wrap) return fail(se, "assertion failed on exception <%s>: %s", signal.getThrowable(), e.getMessage());
+					if (wrap) return errorFormatter.failOptional(se, "assertion failed on exception <%s>: %s", signal.getThrowable(), e.getMessage());
 					throw e;
 				}
 			}
@@ -173,7 +172,7 @@ final class DefaultStepVerifierBuilder<T>
 		checkPotentialHang(1, description);
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnNext()) {
-				return fail(se, "expected: onNext(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onNext(); actual: %s", signal);
 			}
 			else {
 				consumer.accept(signal.get());
@@ -188,7 +187,7 @@ final class DefaultStepVerifierBuilder<T>
 	public DefaultStepVerifierBuilder<T> consumeRecordedWith(
 			Consumer<? super Collection<T>> consumer) {
 		Objects.requireNonNull(consumer, "consumer");
-		this.script.add(new CollectEvent<>(consumer, "consumeRecordedWith"));
+		this.script.add(new CollectEvent<>(consumer, errorFormatter, "consumeRecordedWith"));
 		return this;
 	}
 
@@ -196,10 +195,10 @@ final class DefaultStepVerifierBuilder<T>
 	public DefaultStepVerifierBuilder<T> consumeSubscriptionWith(
 			Consumer<? super Subscription> consumer) {
 		Objects.requireNonNull(consumer, "consumer");
-		if(script.isEmpty() || (script.size() == 1 && script.get(0) == DEFAULT_ONSUBSCRIBE_STEP)) {
+		if(script.isEmpty() || (script.size() == 1 && script.get(0) == defaultFirstStep)) {
 			this.script.set(0, new SignalEvent<>((signal, se) -> {
 				if (!signal.isOnSubscribe()) {
-					return fail(se, "expected: onSubscribe(); actual: %s", signal);
+					return errorFormatter.failOptional(se, "expected: onSubscribe(); actual: %s", signal);
 				}
 				else {
 					consumer.accept(signal.getSubscription());
@@ -216,7 +215,7 @@ final class DefaultStepVerifierBuilder<T>
 
 	@Override
 	public StepVerifier.ContextExpectations<T> expectAccessibleContext() {
-		return new DefaultContextExpectations<>(this);
+		return new DefaultContextExpectations<>(this, errorFormatter);
 	}
 
 	@Override
@@ -231,7 +230,7 @@ final class DefaultStepVerifierBuilder<T>
 					                .blockLast();
 
 					if (c != null) {
-						throw new AssertionError("Expected no accessible Context, got " + c);
+						throw errorFormatter.assertionError("Expected no accessible Context, got " + c);
 					}
 				});
 	}
@@ -240,7 +239,7 @@ final class DefaultStepVerifierBuilder<T>
 	public DefaultStepVerifier<T> expectComplete() {
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnComplete()) {
-				return fail(se, "expected: onComplete(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onComplete(); actual: %s", signal);
 			}
 			else {
 				return Optional.empty();
@@ -254,7 +253,7 @@ final class DefaultStepVerifierBuilder<T>
 	public DefaultStepVerifier<T> expectError() {
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnError()) {
-				return fail(se, "expected: onError(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onError(); actual: %s", signal);
 			}
 			else {
 				return Optional.empty();
@@ -270,11 +269,11 @@ final class DefaultStepVerifierBuilder<T>
 		Objects.requireNonNull(clazz, "clazz");
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnError()) {
-				return fail(se, "expected: onError(%s); actual: %s",
+				return errorFormatter.failOptional(se, "expected: onError(%s); actual: %s",
 						clazz.getSimpleName(), signal);
 			}
 			else if (!clazz.isInstance(signal.getThrowable())) {
-				return fail(se, "expected error of type: %s; actual type: %s",
+				return errorFormatter.failOptional(se, "expected error of type: %s; actual type: %s",
 						clazz.getSimpleName(), signal.getThrowable());
 			}
 			else {
@@ -289,13 +288,13 @@ final class DefaultStepVerifierBuilder<T>
 	public DefaultStepVerifier<T> expectErrorMessage(String errorMessage) {
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnError()) {
-				return fail(se, "expected: onError(\"%s\"); actual: %s",
+				return errorFormatter.failOptional(se, "expected: onError(\"%s\"); actual: %s",
 						errorMessage, signal);
 			}
 			else if (!Objects.equals(errorMessage,
 					signal.getThrowable()
 					      .getMessage())) {
-				return fail(se, "expected error message: \"%s\"; " + "actual " + "message: %s",
+				return errorFormatter.failOptional(se, "expected error message: \"%s\"; " + "actual " + "message: %s",
 						errorMessage,
 						signal.getThrowable()
 						      .getMessage());
@@ -313,10 +312,10 @@ final class DefaultStepVerifierBuilder<T>
 		Objects.requireNonNull(predicate, "predicate");
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnError()) {
-				return fail(se, "expected: onError(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onError(); actual: %s", signal);
 			}
 			else if (!predicate.test(signal.getThrowable())) {
-				return fail(se, "predicate failed on exception: %s", signal.getThrowable());
+				return errorFormatter.failOptional(se, "predicate failed on exception: %s", signal.getThrowable());
 			}
 			else {
 				return Optional.empty();
@@ -405,10 +404,10 @@ final class DefaultStepVerifierBuilder<T>
 		checkPotentialHang(1, desc);
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnNext()) {
-				return fail(se, "expected: onNext(%s); actual: %s", value, signal);
+				return errorFormatter.failOptional(se, "expected: onNext(%s); actual: %s", value, signal);
 			}
 			else if (!Objects.equals(value, signal.get())) {
-				return fail(se, "expected value: %s; actual value: %s", value, signal.get());
+				return errorFormatter.failOptional(se, "expected value: %s; actual value: %s", value, signal.get());
 			}
 			else {
 				return Optional.empty();
@@ -428,7 +427,7 @@ final class DefaultStepVerifierBuilder<T>
 				//best effort
 				checkPotentialHang(-1, "expectNextSequence");
 			}
-			this.script.add(new SignalSequenceEvent<>(iterable, "expectNextSequence"));
+			this.script.add(new SignalSequenceEvent<>(iterable, errorFormatter, "expectNextSequence"));
 		}
 		return this;
 	}
@@ -451,10 +450,10 @@ final class DefaultStepVerifierBuilder<T>
 		checkPotentialHang(1, "expectNextMatches");
 		SignalEvent<T> event = new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnNext()) {
-				return fail(se, "expected: onNext(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onNext(); actual: %s", signal);
 			}
 			else if (!predicate.test(signal.get())) {
-				return fail(se, "predicate failed on value: %s", signal.get());
+				return errorFormatter.failOptional(se, "predicate failed on value: %s", signal.get());
 			}
 			else {
 				return Optional.empty();
@@ -468,17 +467,17 @@ final class DefaultStepVerifierBuilder<T>
 	public DefaultStepVerifierBuilder<T> expectRecordedMatches(
 			Predicate<? super Collection<T>> predicate) {
 		Objects.requireNonNull(predicate, "predicate");
-		this.script.add(new CollectEvent<>(predicate, "expectRecordedMatches"));
+		this.script.add(new CollectEvent<>(predicate, errorFormatter, "expectRecordedMatches"));
 		return this;
 	}
 
 	@Override
 	public DefaultStepVerifierBuilder<T> expectSubscription() {
 		if(this.script.get(0) instanceof NoEvent) {
-			this.script.add(defaultFirstStep());
+			this.script.add(defaultFirstStep);
 		}
 		else{
-			this.script.set(0, newOnSubscribeStep("expectSubscription"));
+			this.script.set(0, newOnSubscribeStep(errorFormatter, "expectSubscription"));
 		}
 		return this;
 	}
@@ -489,10 +488,10 @@ final class DefaultStepVerifierBuilder<T>
 		Objects.requireNonNull(predicate, "predicate");
 		this.script.set(0, new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnSubscribe()) {
-				return fail(se, "expected: onSubscribe(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onSubscribe(); actual: %s", signal);
 			}
 			else if (!predicate.test(signal.getSubscription())) {
-				return fail(se, "predicate failed on subscription: %s",
+				return errorFormatter.failOptional(se, "predicate failed on subscription: %s",
 						signal.getSubscription());
 			}
 			else {
@@ -505,7 +504,7 @@ final class DefaultStepVerifierBuilder<T>
 	@Override
 	public DefaultStepVerifierBuilder<T> expectNoEvent(Duration duration) {
 		Objects.requireNonNull(duration, "duration");
-		if(this.script.size() == 1 && this.script.get(0) == defaultFirstStep()){
+		if(this.script.size() == 1 && this.script.get(0) == defaultFirstStep){
 			this.script.set(0, new NoEvent<>(duration, "expectNoEvent"));
 		}
 		else {
@@ -517,7 +516,7 @@ final class DefaultStepVerifierBuilder<T>
 	@Override
 	public DefaultStepVerifierBuilder<T> recordWith(Supplier<? extends Collection<T>> supplier) {
 		Objects.requireNonNull(supplier, "supplier");
-		this.script.add(new CollectEvent<>(supplier, "recordWith"));
+		this.script.add(new CollectEvent<>(supplier, errorFormatter, "recordWith"));
 		return this;
 	}
 
@@ -623,7 +622,7 @@ final class DefaultStepVerifierBuilder<T>
 			} else {
 				message.append(expectedAmount);
 			}
-			throw new IllegalArgumentException(message.toString());
+			throw errorFormatter.error(IllegalArgumentException::new, message.toString());
 		}
 		else {
 			this.hangCheckRequested -= expectedAmount;
@@ -680,7 +679,7 @@ final class DefaultStepVerifierBuilder<T>
 				Duration time = verify();
 
 				//return the assertion API
-				return new DefaultStepVerifierAssertions(droppedElements, droppedErrors, operatorErrors, time);
+				return new DefaultStepVerifierAssertions(droppedElements, droppedErrors, operatorErrors, time, parent.errorFormatter);
 			}
 			finally {
 				//unplug the hooks
@@ -712,6 +711,7 @@ final class DefaultStepVerifierBuilder<T>
 
 					DefaultVerifySubscriber<T> newVerifier = new DefaultVerifySubscriber<>(
 							this.parent.script,
+							this.parent.errorFormatter,
 							this.parent.initialRequest,
 							this.requestedFusionMode,
 							this.expectedFusionMode,
@@ -758,6 +758,7 @@ final class DefaultStepVerifierBuilder<T>
 			}
 			return new DefaultVerifySubscriber<>(
 					this.parent.script,
+					this.parent.errorFormatter,
 					this.parent.initialRequest,
 					this.requestedFusionMode,
 					this.expectedFusionMode,
@@ -774,6 +775,7 @@ final class DefaultStepVerifierBuilder<T>
 
 		final CountDownLatch                completeLatch;
 		final Queue<Event<T>>               script;
+		final ErrorFormatter                errorFormatter;
 		final Queue<TaskEvent<T>>           taskEvents;
 		final int                           requestedFusionMode;
 		final int                           expectedFusionMode;
@@ -781,6 +783,7 @@ final class DefaultStepVerifierBuilder<T>
 		final Context                       initialContext;
 		final VirtualTimeScheduler          virtualTimeScheduler;
 
+		@Nullable
 		Logger                        logger;
 		int                           establishedFusionMode;
 		Fuseable.QueueSubscription<T> qs;
@@ -806,6 +809,7 @@ final class DefaultStepVerifierBuilder<T>
 		 * passed */
 		@SuppressWarnings("unchecked")
 		DefaultVerifySubscriber(List<Event<T>> script,
+				ErrorFormatter errorFormatter,
 				long initialRequest,
 				int requestedFusionMode,
 				int expectedFusionMode,
@@ -835,6 +839,7 @@ final class DefaultStepVerifierBuilder<T>
 			this.completeLatch = new CountDownLatch(1);
 			this.requested = initialRequest;
 			this.initialContext = initialContext == null ? Context.empty() : initialContext;
+			this.errorFormatter = errorFormatter;
 		}
 
 		static <R> Queue<Event<R>> conflateScript(List<Event<R>> script, @Nullable Logger logger) {
@@ -881,7 +886,7 @@ final class DefaultStepVerifierBuilder<T>
 					logger.debug("\t<{}>", current.getDescription());
 				}
 			}
-			//TODO simplified whole algo, remove DescriptionTasks
+			//TODO simplify whole algo, remove DescriptionTasks
 
 			return conflated;
 		}
@@ -1077,7 +1082,7 @@ final class DefaultStepVerifierBuilder<T>
 				Duration time = verify();
 
 				//return the assertion API
-				return new DefaultStepVerifierAssertions(droppedElements, droppedErrors, operatorErrors, time);
+				return new DefaultStepVerifierAssertions(droppedElements, droppedErrors, operatorErrors, time, errorFormatter);
 			}
 			finally {
 				//unplug the hooks
@@ -1131,13 +1136,13 @@ final class DefaultStepVerifierBuilder<T>
 		 * @param arguments the optional formatter arguments to the message
 		 */
 		final void setFailure(@Nullable Event<T> event, @Nullable Signal<T> actualSignal, String msg, Object... arguments) {
-			Exceptions.addThrowable(ERRORS, this, fail(event, msg, arguments).get());
+			Exceptions.addThrowable(ERRORS, this, errorFormatter.fail(event, msg, arguments));
 			maybeCancel(actualSignal);
 			this.completeLatch.countDown();
 		}
 
 		final void setFailurePrefix(String prefix, Signal<T> actualSignal, String msg, Object... arguments) {
-			Exceptions.addThrowable(ERRORS, this, failPrefix(prefix, msg, arguments).get());
+			Exceptions.addThrowable(ERRORS, this, errorFormatter.failPrefix(prefix, msg, arguments));
 			maybeCancel(actualSignal);
 			this.completeLatch.countDown();
 		}
@@ -1165,8 +1170,7 @@ final class DefaultStepVerifierBuilder<T>
 		final Optional<AssertionError> checkCountMismatch(SignalCountEvent<T> event, Signal<T> s) {
 			long expected = event.count;
 			if (!s.isOnNext()) {
-				return fail(event, "expected: count = %s; actual: " + "counted = %s; " +
-								"signal: %s",
+				return errorFormatter.failOptional(event, "expected: count = %s; actual: counted = %s; signal: %s",
 						expected,
 						unasserted, s);
 			}
@@ -1299,7 +1303,7 @@ final class DefaultStepVerifierBuilder<T>
 				}
 				else {
 					String msg = e.getMessage() != null ? e.getMessage() : "";
-					AssertionError wrapFailure = fail(null,
+					AssertionError wrapFailure = errorFormatter.failOptional(null,
 							"failed running expectation on signal [%s] " + "with " + "[%s]:\n%s",
 							actualSignal,
 							Exceptions.unwrap(e)
@@ -1507,11 +1511,10 @@ final class DefaultStepVerifierBuilder<T>
 				}
 				if (timeout != Duration.ZERO && stop.isBefore(Instant.now())) {
 					if (get() == null) {
-						throw new IllegalStateException(
-								"VerifySubscriber has not been subscribed");
+						throw errorFormatter.error(IllegalStateException::new, "VerifySubscriber has not been subscribed");
 					}
 					else {
-						throw new AssertionError("VerifySubscriber timed out on " + get());
+						throw errorFormatter.assertionError("VerifySubscriber timed out on " + get());
 					}
 				}
 			}
@@ -1610,8 +1613,7 @@ final class DefaultStepVerifierBuilder<T>
 		@SuppressWarnings("unchecked")
 		final void validate() {
 			if (get() == null) {
-				throw new IllegalStateException(
-						"VerifySubscriber has not been subscribed");
+				throw errorFormatter.error(IllegalStateException::new, "VerifySubscriber has not been subscribed");
 			}
 			Throwable errors = this.errors;
 
@@ -1633,7 +1635,7 @@ final class DefaultStepVerifierBuilder<T>
 			             .forEach(messageBuilder::append);
 
 			messageBuilder.delete(messageBuilder.length() - 1, messageBuilder.length());
-			throw new AssertionError(messageBuilder.toString(), errors);
+			throw errorFormatter.assertionError(messageBuilder.toString(), errors);
 		}
 
 	}
@@ -1644,20 +1646,23 @@ final class DefaultStepVerifierBuilder<T>
 		private final Queue<Throwable> droppedErrors;
 		private final Queue<Tuple2<Optional<Throwable>, Optional<?>>> operatorErrors;
 		private final Duration duration;
+		private final ErrorFormatter errorFormatter;
 
 		DefaultStepVerifierAssertions(Queue<Object> droppedElements,
 				Queue<Throwable> droppedErrors,
 				Queue<Tuple2<Optional<Throwable>, Optional<?>>> operatorErrors,
-				Duration duration) {
+				Duration duration,
+				ErrorFormatter errorFormatter) {
 			this.droppedElements = droppedElements;
 			this.droppedErrors = droppedErrors;
 			this.operatorErrors = operatorErrors;
 			this.duration = duration;
+			this.errorFormatter = errorFormatter;
 		}
 
 		private StepVerifier.Assertions satisfies(BooleanSupplier check, Supplier<String> message) {
 			if (!check.getAsBoolean()) {
-				throw new AssertionError(message.get());
+				throw errorFormatter.assertionError(message.get());
 			}
 			return this;
 		}
@@ -1980,28 +1985,33 @@ final class DefaultStepVerifierBuilder<T>
 
 	static final class CollectEvent<T> extends AbstractEagerEvent<T> {
 
+		final ErrorFormatter errorFormatter;
+
 		final Supplier<? extends Collection<T>> supplier;
 
 		final Predicate<? super Collection<T>>  predicate;
 
 		final Consumer<? super Collection<T>>   consumer;
 
-		CollectEvent(Supplier<? extends Collection<T>> supplier, String desc) {
+		CollectEvent(Supplier<? extends Collection<T>> supplier, ErrorFormatter errorFormatter, String desc) {
 			super(desc);
+			this.errorFormatter = errorFormatter;
 			this.supplier = supplier;
 			this.predicate = null;
 			this.consumer = null;
 		}
 
-		CollectEvent(Consumer<? super Collection<T>> consumer, String desc) {
+		CollectEvent(Consumer<? super Collection<T>> consumer, ErrorFormatter errorFormatter, String desc) {
 			super(desc);
+			this.errorFormatter = errorFormatter;
 			this.supplier = null;
 			this.predicate = null;
 			this.consumer = consumer;
 		}
 
-		CollectEvent(Predicate<? super Collection<T>> predicate, String desc) {
+		CollectEvent(Predicate<? super Collection<T>> predicate, ErrorFormatter errorFormatter, String desc) {
 			super(desc);
+			this.errorFormatter = errorFormatter;
 			this.supplier = null;
 			this.predicate = predicate;
 			this.consumer = null;
@@ -2015,7 +2025,7 @@ final class DefaultStepVerifierBuilder<T>
 		Optional<AssertionError> test(Collection<T> collection) {
 			if (predicate != null) {
 				if (!predicate.test(collection)) {
-					return fail(this, "expected collection predicate match; actual: %s",
+					return errorFormatter.failOptional(this, "expected collection predicate match; actual: %s",
 							collection);
 				}
 				else {
@@ -2087,7 +2097,7 @@ final class DefaultStepVerifierBuilder<T>
 				virtualOrRealWait(duration.minus(Duration.ofNanos(1)), parent);
 				parent.monitorSignal = false;
 				if(parent.isTerminated() && !parent.isCancelled()){
-					throw new AssertionError("unexpected end during a no-event expectation");
+					throw parent.errorFormatter.assertionError("unexpected end during a no-event expectation");
 				}
 				virtualOrRealWait(Duration.ofNanos(1), parent);
 			}
@@ -2096,7 +2106,7 @@ final class DefaultStepVerifierBuilder<T>
 				virtualOrRealWait(duration, parent);
 				parent.monitorSignal = false;
 				if(parent.isTerminated() && !parent.isCancelled()){
-					throw new AssertionError("unexpected end during a no-event expectation");
+					throw parent.errorFormatter.assertionError("unexpected end during a no-event expectation");
 				}
 			}
 		}
@@ -2144,10 +2154,12 @@ final class DefaultStepVerifierBuilder<T>
 	static final class SignalSequenceEvent<T> extends AbstractSignalEvent<T> {
 
 		final Iterable<? extends T> iterable;
+		final ErrorFormatter        errorFormatter;
 
-		SignalSequenceEvent(Iterable<? extends T> iterable, String desc) {
+		SignalSequenceEvent(Iterable<? extends T> iterable, ErrorFormatter errorFormatter, String desc) {
 			super(desc);
 			this.iterable = iterable;
+			this.errorFormatter = errorFormatter;
 		}
 
 		Optional<AssertionError> test(Signal<T> signal, Iterator<? extends T> iterator) {
@@ -2157,7 +2169,7 @@ final class DefaultStepVerifierBuilder<T>
 				}
 				T d2 = iterator.next();
 				if (!Objects.equals(signal.get(), d2)) {
-					return fail(this, "expected : onNext(%s); actual: %s; iterable: %s",
+					return errorFormatter.failOptional(this, "expected : onNext(%s); actual: %s; iterable: %s",
 							d2,
 							signal.get(),
 							iterable);
@@ -2166,7 +2178,7 @@ final class DefaultStepVerifierBuilder<T>
 
 			}
 			if (iterator.hasNext() || signal.isOnError()) {
-				return fail(this, "expected next value: %s; actual signal: %s; iterable: %s",
+				return errorFormatter.failOptional(this, "expected next value: %s; actual signal: %s; iterable: %s",
 						iterator.hasNext() ? iterator.next() : "none",
 						signal, iterable);
 			}
@@ -2214,19 +2226,6 @@ final class DefaultStepVerifierBuilder<T>
 		}
 	}
 
-	static Optional<AssertionError> fail(@Nullable Event<?> event, String msg, Object... args) {
-		String prefix = "expectation failed (";
-		if (event != null && event.getDescription().length() > 0) {
-			prefix = String.format("expectation \"%s\" failed (", event.getDescription());
-		}
-
-		return failPrefix(prefix, msg, args);
-	}
-
-	static Optional<AssertionError> failPrefix(String prefix, String msg, Object... args) {
-		return Optional.of(new AssertionError(prefix + String.format(msg, args) + ")"));
-	}
-
 	static String formatFusionMode(int m) {
 		switch (m) {
 			case Fuseable.ANY:
@@ -2243,10 +2242,10 @@ final class DefaultStepVerifierBuilder<T>
 		return "" + m;
 	}
 
-	static <T> SignalEvent<T> newOnSubscribeStep(String desc){
+	static <T> SignalEvent<T> newOnSubscribeStep(ErrorFormatter errorFormatter, String desc){
 		return new SignalEvent<>((signal, se) -> {
 			if (!signal.isOnSubscribe()) {
-				return fail(se, "expected: onSubscribe(); actual: %s", signal);
+				return errorFormatter.failOptional(se, "expected: onSubscribe(); actual: %s", signal);
 			}
 			else {
 				return Optional.empty();
@@ -2257,13 +2256,15 @@ final class DefaultStepVerifierBuilder<T>
 	static final class DefaultContextExpectations<T>
 			implements StepVerifier.ContextExpectations<T> {
 
+		private final ErrorFormatter       errorFormatter;
 		private final StepVerifier.Step<T> step;
 		private Consumer<Context>          contextExpectations;
 
-		DefaultContextExpectations(StepVerifier.Step<T> step) {
+		DefaultContextExpectations(StepVerifier.Step<T> step, ErrorFormatter errorFormatter) {
+			this.errorFormatter = errorFormatter;
 			this.step = step;
 			this.contextExpectations = c -> {
-				if (c == null) throw new AssertionError("No propagated Context");
+				if (c == null) throw errorFormatter.assertionError("No propagated Context");
 			};
 		}
 
@@ -2286,8 +2287,7 @@ final class DefaultStepVerifierBuilder<T>
 		public StepVerifier.ContextExpectations<T> hasKey(Object key) {
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 					if (!c.hasKey(key))
-						throw new AssertionError(String.format(
-								"Key %s not found in Context %s", key, c));
+						throw errorFormatter.assertionError(String.format("Key %s not found in Context %s", key, c));
 			});
 			return this;
 		}
@@ -2297,8 +2297,8 @@ final class DefaultStepVerifierBuilder<T>
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				long realSize = c.stream().count();
 				if (realSize != size)
-					throw new AssertionError(String.format(
-							"Expected Context of size %d, got %d for Context %s", size, realSize, c));
+					throw errorFormatter.assertionError(
+							String.format("Expected Context of size %d, got %d for Context %s", size, realSize, c));
 			});
 			return this;
 		}
@@ -2308,12 +2308,12 @@ final class DefaultStepVerifierBuilder<T>
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				Object realValue = c.getOrDefault(key, null);
 				if (realValue == null)
-					throw new AssertionError(String.format(
-							"Expected value %s for key %s, key not present in Context %s", value, key, c));
+					throw errorFormatter.assertionError(
+							String.format("Expected value %s for key %s, key not present in Context %s", value, key, c));
 
 				if (!value.equals(realValue))
-					throw new AssertionError(String.format(
-							"Expected value %s for key %s, got %s in Context %s", value, key, realValue, c));
+					throw errorFormatter.assertionError(
+							String.format("Expected value %s for key %s, got %s in Context %s", value, key, realValue, c));
 			});
 			return this;
 		}
@@ -2323,7 +2323,7 @@ final class DefaultStepVerifierBuilder<T>
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				boolean all = other.stream().allMatch(e -> e.getValue().equals(c.getOrDefault(e.getKey(), null)));
 				if (!all) {
-					throw new AssertionError(String.format("Expected Context %s to contain all of %s", c, other));
+					throw errorFormatter.assertionError(String.format("Expected Context %s to contain all of %s", c, other));
 				}
 			});
 			return this;
@@ -2336,7 +2336,7 @@ final class DefaultStepVerifierBuilder<T>
 				                   .stream()
 				                   .allMatch(e -> e.getValue().equals(c.getOrDefault(e.getKey(), null)));
 				if (!all) {
-					throw new AssertionError(String.format("Expected Context %s to contain all of %s", c, other));
+					throw errorFormatter.assertionError(String.format("Expected Context %s to contain all of %s", c, other));
 				}
 			});
 			return this;
@@ -2346,12 +2346,14 @@ final class DefaultStepVerifierBuilder<T>
 		public StepVerifier.ContextExpectations<T> containsOnly(Context other) {
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				if (c.stream().count() != other.stream().count()) {
-					throw new AssertionError(String.format("Expected Context %s to contain same values as %s, but they differ in size", c, other));
+					throw errorFormatter.assertionError(
+							String.format("Expected Context %s to contain same values as %s, but they differ in size", c, other));
 				}
 				boolean all = other.stream()
 				                   .allMatch(e -> e.getValue().equals(c.getOrDefault(e.getKey(), null)));
 				if (!all) {
-					throw new AssertionError(String.format("Expected Context %s to contain same values as %s, but they differ in content", c, other));
+					throw errorFormatter.assertionError(
+							String.format("Expected Context %s to contain same values as %s, but they differ in content", c, other));
 				}
 			});
 			return this;
@@ -2361,13 +2363,15 @@ final class DefaultStepVerifierBuilder<T>
 		public StepVerifier.ContextExpectations<T> containsOnly(Map<?, ?> other) {
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				if (c.stream().count() != other.size()) {
-					throw new AssertionError(String.format("Expected Context %s to contain same values as %s, but they differ in size", c, other));
+					throw errorFormatter.assertionError(
+							String.format("Expected Context %s to contain same values as %s, but they differ in size", c, other));
 				}
 				boolean all = other.entrySet()
 				                   .stream()
 				                   .allMatch(e -> e.getValue().equals(c.getOrDefault(e.getKey(), null)));
 				if (!all) {
-					throw new AssertionError(String.format("Expected Context %s to contain same values as %s, but they differ in content", c, other));
+					throw errorFormatter.assertionError(
+							String.format("Expected Context %s to contain same values as %s, but they differ in content", c, other));
 				}
 			});
 			return this;
@@ -2383,7 +2387,8 @@ final class DefaultStepVerifierBuilder<T>
 		public StepVerifier.ContextExpectations<T> matches(Predicate<Context> predicate) {
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				if (!predicate.test(c)) {
-					throw new AssertionError(String.format("Context %s doesn't match predicate", c));
+					throw errorFormatter.assertionError(
+							String.format("Context %s doesn't match predicate", c));
 				}
 			});
 			return this;
@@ -2393,14 +2398,13 @@ final class DefaultStepVerifierBuilder<T>
 		public StepVerifier.ContextExpectations<T> matches(Predicate<Context> predicate, String description) {
 			this.contextExpectations = this.contextExpectations.andThen(c -> {
 				if (!predicate.test(c)) {
-					throw new AssertionError(String.format("Context %s doesn't match predicate %s", c, description));
+					throw errorFormatter.assertionError(
+							String.format("Context %s doesn't match predicate %s", c, description));
 				}
 			});
 			return this;
 		}
 	}
-
-	static final SignalEvent DEFAULT_ONSUBSCRIBE_STEP = newOnSubscribeStep("defaultOnSubscribe");
 
 	static final AtomicReferenceFieldUpdater<DefaultVerifySubscriber, Throwable>
 			ERRORS =

--- a/reactor-test/src/main/java/reactor/test/ErrorFormatter.java
+++ b/reactor-test/src/main/java/reactor/test/ErrorFormatter.java
@@ -1,0 +1,131 @@
+/*
+ * Copyright (c) 2011-2018 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.test;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import reactor.util.annotation.Nullable;
+
+/**
+ * A utility class to format error messages in all of the default implementation of {@link StepVerifier}.
+ * Formats events and is capable of prepending messages with a {@link StepVerifierOptions#scenarioName(String) scenario name}.
+ *
+ * @author Simon Baslé
+ */
+final class ErrorFormatter {
+
+	final String scenarioPrefix;
+
+	ErrorFormatter(@Nullable final String scenarioName) {
+		if (scenarioName == null || scenarioName.isEmpty()) {
+			scenarioPrefix = "";
+		}
+		else {
+			scenarioPrefix = "[" + scenarioName + "] ";
+		}
+	}
+
+	/**
+	 * Format an event, message with argument placeholders (in the style of {@link String#format(String, Object...)})
+	 * and produce an {@link AssertionError}.
+	 *
+	 * @param event the event that caused an expectation failure
+	 * @param msg the message, possibly with {@link String#format(String, Object...)} style placeholders
+	 * @param args the optional values for the placeholders in msg
+	 * @return an {@link AssertionError} with a standardized message potentially prefixed with the associated scenario name
+	 */
+	AssertionError fail(@Nullable DefaultStepVerifierBuilder.Event<?> event, String msg, Object... args) {
+		String prefix;
+		if (event != null && event.getDescription()
+		                          .length() > 0) {
+			prefix = String.format("expectation \"%s\" failed (", event.getDescription());
+		}
+		else {
+			prefix = "expectation failed (";
+		}
+
+		return failPrefix(prefix, msg, args);
+	}
+
+	/**
+	 * Format an event, message with argument placeholders (in the style of {@link String#format(String, Object...)})
+	 * and produce an {@link AssertionError} wrapped in an {@link Optional}.
+	 *
+	 * @param event the event that caused an expectation failure
+	 * @param msg the message, possibly with {@link String#format(String, Object...)} style placeholders
+	 * @param args the optional values for the placeholders in msg
+	 * @return an {@link AssertionError} with a standardized message potentially prefixed with the associated scenario name,
+	 * wrapped in an {@link Optional}
+	 */
+	Optional<AssertionError> failOptional(@Nullable DefaultStepVerifierBuilder.Event<?> event, String msg,
+			Object... args) {
+		return Optional.of(fail(event, msg, args));
+	}
+
+	/**
+	 * Formats a two-part message: a prefix that should end in an opening parenthesis and
+	 * a second part with argument placeholders (in the style of {@link String#format(String, Object...)}).
+	 * This method adds the closing parenthesis after that second part, and then produces
+	 * an {@link AssertionError} out of the formatted message.
+	 *
+	 * @param prefix the first part of the message, should en with an opening parenthesis
+	 * @param msg the second part (detail) of the message, possibly with {@link String#format(String, Object...)} style placeholders
+	 * @param args the optional values for the placeholders in msg
+	 * @return an {@link AssertionError} with a standardized message potentially prefixed with the associated scenario name
+	 */
+	AssertionError failPrefix(String prefix, String msg, Object... args) {
+		return assertionError(prefix + String.format(msg, args) + ")");
+	}
+
+	/**
+	 * Produce an {@link AssertionError} out of a given plain message, potentially prefixing
+	 * it with the associated scenario name.
+	 *
+	 * @param msg the plain message
+	 * @return an {@link AssertionError} with a standardized message potentially prefixed with the associated scenario name
+	 */
+	AssertionError assertionError(String msg) {
+		return new AssertionError(scenarioPrefix + msg);
+	}
+
+	/**
+	 * Produce an {@link AssertionError} out of a given plain message and a cause,
+	 * potentially prefixing it with the associated scenario name.
+	 *
+	 * @param msg the plain message
+	 * @param cause the cause to add to the {@link AssertionError}
+	 * @return an {@link AssertionError} with a cause and a standardized message
+	 * potentially prefixed with the associated scenario name
+	 */
+	AssertionError assertionError(String msg, Throwable cause) {
+		return new AssertionError(scenarioPrefix + msg, cause);
+	}
+
+	/**
+	 * Produce an arbitrary {@link Throwable} with a standardized message comprised of
+	 * the given plain message and an optional prefix if the associated scenario is named.
+	 *
+	 * @param errorProducer the error producer (usually an Exception constructor method reference)
+	 * @param message the plain message
+	 * @param <T> the type of the {@link Exception}
+	 * @return a {@link T} with a message potentially prefixed with the associated scenario name
+	 */
+	<T extends Throwable> T error(Function<String, T> errorProducer, String message) {
+		return errorProducer.apply(scenarioPrefix + message);
+	}
+}

--- a/reactor-test/src/main/java/reactor/test/ErrorFormatter.java
+++ b/reactor-test/src/main/java/reactor/test/ErrorFormatter.java
@@ -29,11 +29,13 @@ import reactor.util.annotation.Nullable;
  */
 final class ErrorFormatter {
 
+	private static final String EMPTY = "";
+
 	final String scenarioPrefix;
 
 	ErrorFormatter(@Nullable final String scenarioName) {
 		if (scenarioName == null || scenarioName.isEmpty()) {
-			scenarioPrefix = "";
+			scenarioPrefix = EMPTY;
 		}
 		else {
 			scenarioPrefix = "[" + scenarioName + "] ";

--- a/reactor-test/src/main/java/reactor/test/StepVerifierOptions.java
+++ b/reactor-test/src/main/java/reactor/test/StepVerifierOptions.java
@@ -29,6 +29,9 @@ import reactor.util.context.Context;
  */
 public class StepVerifierOptions {
 
+	@Nullable
+	private String scenarioName = null;
+
 	private boolean checkUnderRequesting = true;
 	private long initialRequest = Long.MAX_VALUE;
 	private Supplier<? extends VirtualTimeScheduler> vtsLookup = null;
@@ -124,5 +127,14 @@ public class StepVerifierOptions {
 	@Nullable
 	public Context getInitialContext() {
 		return this.initialContext;
+	}
+
+	public StepVerifierOptions scenarioName(@Nullable String scenarioName) {
+		this.scenarioName = scenarioName;
+		return this;
+	}
+
+	public String getScenarioName() {
+		return this.scenarioName;
 	}
 }

--- a/reactor-test/src/main/java/reactor/test/StepVerifierOptions.java
+++ b/reactor-test/src/main/java/reactor/test/StepVerifierOptions.java
@@ -129,11 +129,24 @@ public class StepVerifierOptions {
 		return this.initialContext;
 	}
 
+	/**
+	 * Give a name to the whole scenario tested by the configured {@link StepVerifier}. That
+	 * name would be mentioned in exceptions and assertion errors raised by the StepVerifier,
+	 * allowing to better distinguish error sources in unit tests where multiple StepVerifier
+	 * are used.
+	 *
+	 * @param scenarioName the name of the scenario, null to deactivate
+	 * @return this instance, to continue setting the options.
+	 */
 	public StepVerifierOptions scenarioName(@Nullable String scenarioName) {
 		this.scenarioName = scenarioName;
 		return this;
 	}
 
+	/**
+	 * @return the name given to the configured {@link StepVerifier}, or null if none.
+	 */
+	@Nullable
 	public String getScenarioName() {
 		return this.scenarioName;
 	}

--- a/reactor-test/src/test/java/reactor/test/DefaultContextExpectationsTest.java
+++ b/reactor-test/src/test/java/reactor/test/DefaultContextExpectationsTest.java
@@ -41,7 +41,7 @@ public class DefaultContextExpectationsTest {
 			long count) {
 		Flux<Integer> source = sourceTransformer.apply(Flux.range(1, 10));
 		Step<Integer> step = StepVerifier.create(source);
-		final DefaultContextExpectations<Integer> base = new DefaultContextExpectations<>(step);
+		final DefaultContextExpectations<Integer> base = new DefaultContextExpectations<>(step, new ErrorFormatter(null));
 
 		expectations
 				.apply(base)
@@ -259,6 +259,22 @@ public class DefaultContextExpectationsTest {
 		assertContextExpectationFails(s -> s.subscriberContext(Context.of("foo", "bar")),
 				e -> e.matches(Objects::isNull, "desc"))
 				.withMessage("Context Context1{foo=bar} doesn't match predicate desc");
+	}
+
+	@Test
+	public void notMatchesWithDescriptionAndScenarioName() {
+		Flux<Integer> source = Flux.range(1, 10)
+		                           .subscriberContext(Context.of("foo", "bar"));
+
+		Step<Integer> step = StepVerifier.create(source);
+		final DefaultContextExpectations<Integer> base = new DefaultContextExpectations<>(step, new ErrorFormatter("scenario"));
+
+		assertThatExceptionOfType(AssertionError.class)
+				.isThrownBy(
+						base.matches(Objects::isNull, "someDescription")
+						    .then()
+						    .expectNextCount(10)::verifyComplete)
+				.withMessage("[scenario] Context Context1{foo=bar} doesn't match predicate someDescription");
 	}
 
 }

--- a/reactor-test/src/test/java/reactor/test/ErrorFormatterTest.java
+++ b/reactor-test/src/test/java/reactor/test/ErrorFormatterTest.java
@@ -1,0 +1,231 @@
+/*
+ * Copyright (c) 2011-2018 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.test;
+
+import java.time.Duration;
+
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ErrorFormatterTest {
+
+	@Test
+	public void noScenarioEmpty() {
+		assertThat(new ErrorFormatter("").scenarioPrefix)
+				.isNotNull()
+				.isEmpty();
+	}
+
+	@Test
+	public void nullScenarioEmpty() {
+		assertThat(new ErrorFormatter(null).scenarioPrefix)
+				.isNotNull()
+				.isEmpty();
+	}
+
+	@Test
+	public void givenScenarioWrapped() {
+		assertThat(new ErrorFormatter("foo").scenarioPrefix)
+				.isEqualTo("[foo] ");
+	}
+
+	// === Tests with an empty scenario name ===
+	static final ErrorFormatter noScenario = new ErrorFormatter("");
+
+	@Test
+	public void noScenarioFailNullEventNoArgs() {
+		assertThat(noScenario.fail(null, "details"))
+				.hasMessage("expectation failed (details)");
+	}
+
+	@Test
+	public void noScenarioFailNoDescriptionNoArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"");
+
+		assertThat(noScenario.fail(event, "details"))
+				.hasMessage("expectation failed (details)");
+	}
+
+	@Test
+	public void noScenarioFailDescriptionNoArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"eventDescription");
+
+		assertThat(noScenario.fail(event, "details"))
+				.hasMessage("expectation \"eventDescription\" failed (details)");
+	}
+
+	@Test
+	public void noScenarioFailNullEventHasArgs() {
+		assertThat(noScenario.fail(null, "details = %s", "bar"))
+				.hasMessage("expectation failed (details = bar)");
+	}
+
+
+	@Test
+	public void noScenarioFailNoDescriptionHasArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"");
+
+		assertThat(noScenario.fail(event, "details = %s", "bar"))
+				.hasMessage("expectation failed (details = bar)");
+	}
+
+	@Test
+	public void noScenarioFailDescriptionHasArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"eventDescription");
+
+		assertThat(noScenario.fail(event, "details = %s", "bar"))
+				.hasMessage("expectation \"eventDescription\" failed (details = bar)");
+	}
+
+	@Test
+	public void noScenarioFailOptional() {
+		assertThat(noScenario.failOptional(null, "foo"))
+				.hasValueSatisfying(ae -> assertThat(ae).hasMessage("expectation failed (foo)"));
+	}
+
+	@Test
+	public void noScenarioFailPrefixNoArgs() {
+		assertThat(noScenario.failPrefix("firstPart", "secondPart"))
+				.hasMessage("firstPartsecondPart)"); //note the prefix doesn't have an opening parenthesis
+	}
+
+	@Test
+	public void noScenarioFailPrefixHasArgs() {
+		assertThat(noScenario.failPrefix("firstPart(", "secondPart = %s", "foo"))
+				.hasMessage("firstPart(secondPart = foo)");
+	}
+
+	@Test
+	public void noScenarioAssertionError() {
+		assertThat(noScenario.assertionError("plain"))
+				.hasMessage("plain")
+				.hasNoCause();
+	}
+
+	@Test
+	public void noScenarioAssertionErrorWithCause() {
+		Throwable cause = new IllegalArgumentException("boom");
+		assertThat(noScenario.assertionError("plain", cause))
+				.hasMessage("plain")
+				.hasCause(cause);
+	}
+	
+	@Test
+	public void noScenarioIllegalStateException() {
+		assertThat(noScenario.<Throwable>error(IllegalStateException::new, "plain"))
+				.isInstanceOf(IllegalStateException.class)
+				.hasMessage("plain");
+	}
+	
+	
+	// === Tests with a scenario name ===
+	static final ErrorFormatter withScenario = new ErrorFormatter("ErrorFormatterTest");
+
+	@Test
+	public void withScenarioFailNullEventNoArgs() {
+		assertThat(withScenario.fail(null, "details"))
+				.hasMessage("[ErrorFormatterTest] expectation failed (details)");
+	}
+
+	@Test
+	public void withScenarioFailNoDescriptionNoArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"");
+
+		assertThat(withScenario.fail(event, "details"))
+				.hasMessage("[ErrorFormatterTest] expectation failed (details)");
+	}
+
+	@Test
+	public void withScenarioFailDescriptionNoArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"eventDescription");
+
+		assertThat(withScenario.fail(event, "details"))
+				.hasMessage("[ErrorFormatterTest] expectation \"eventDescription\" failed (details)");
+	}
+
+	@Test
+	public void withScenarioFailNullEventHasArgs() {
+		assertThat(withScenario.fail(null, "details = %s", "bar"))
+				.hasMessage("[ErrorFormatterTest] expectation failed (details = bar)");
+	}
+
+
+	@Test
+	public void withScenarioFailNoDescriptionHasArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"");
+
+		assertThat(withScenario.fail(event, "details = %s", "bar"))
+				.hasMessage("[ErrorFormatterTest] expectation failed (details = bar)");
+	}
+
+	@Test
+	public void withScenarioFailDescriptionHasArgs() {
+		DefaultStepVerifierBuilder.Event event = new DefaultStepVerifierBuilder.NoEvent(Duration.ofMillis(5),
+				"eventDescription");
+
+		assertThat(withScenario.fail(event, "details = %s", "bar"))
+				.hasMessage("[ErrorFormatterTest] expectation \"eventDescription\" failed (details = bar)");
+	}
+
+	@Test
+	public void withScenarioFailOptional() {
+		assertThat(withScenario.failOptional(null, "foo"))
+				.hasValueSatisfying(ae -> assertThat(ae).hasMessage("[ErrorFormatterTest] expectation failed (foo)"));
+	}
+
+	@Test
+	public void withScenarioFailPrefixNoArgs() {
+		assertThat(withScenario.failPrefix("firstPart", "secondPart"))
+				.hasMessage("[ErrorFormatterTest] firstPartsecondPart)"); //note the prefix doesn't have an opening parenthesis
+	}
+
+	@Test
+	public void withScenarioFailPrefixHasArgs() {
+		assertThat(withScenario.failPrefix("firstPart(", "secondPart = %s", "foo"))
+				.hasMessage("[ErrorFormatterTest] firstPart(secondPart = foo)");
+	}
+
+	@Test
+	public void withScenarioAssertionError() {
+		assertThat(withScenario.assertionError("plain"))
+				.hasMessage("[ErrorFormatterTest] plain")
+				.hasNoCause();
+	}
+
+	@Test
+	public void withScenarioAssertionErrorWithCause() {
+		Throwable cause = new IllegalArgumentException("boom");
+		assertThat(withScenario.assertionError("plain", cause))
+				.hasMessage("[ErrorFormatterTest] plain")
+				.hasCause(cause);
+	}
+	
+	@Test
+	public void withScenarioIllegalStateException() {
+		assertThat(withScenario.<Throwable>error(IllegalStateException::new, "plain"))
+				.isInstanceOf(IllegalStateException.class)
+				.hasMessage("[ErrorFormatterTest] plain");
+	}
+}

--- a/reactor-test/src/test/java/reactor/test/StepVerifierAssertionsTests.java
+++ b/reactor-test/src/test/java/reactor/test/StepVerifierAssertionsTests.java
@@ -24,6 +24,7 @@ import reactor.core.publisher.Operators;
 import reactor.util.context.Context;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.assertj.core.api.Assertions.fail;
 
 public class StepVerifierAssertionsTests {
@@ -515,9 +516,23 @@ public class StepVerifierAssertionsTests {
 
 	@Test
 	public void assertDurationConsidersEqualsASuccess() {
-		new DefaultStepVerifierBuilder.DefaultStepVerifierAssertions(null, null, null, Duration.ofSeconds(3))
+		new DefaultStepVerifierBuilder.DefaultStepVerifierAssertions(null, null, null,
+				Duration.ofSeconds(3),
+				new ErrorFormatter(null))
 				.tookLessThan(Duration.ofMillis(3000L))
 				.tookMoreThan(Duration.ofSeconds(3));
+	}
+
+	@Test
+	public void assertDurationFailureWithScenarioName() {
+		assertThatExceptionOfType(AssertionError.class)
+				.isThrownBy(() ->
+						new DefaultStepVerifierBuilder.DefaultStepVerifierAssertions(null, null, null
+								, Duration.ofSeconds(3),
+								new ErrorFormatter("fooScenario"))
+								.tookLessThan(Duration.ofMillis(200))
+				)
+				.withMessage("[fooScenario] Expected scenario to be verified in less than 200ms, took 3000ms.");
 	}
 
 	@Test

--- a/reactor-test/src/test/java/reactor/test/StepVerifierTests.java
+++ b/reactor-test/src/test/java/reactor/test/StepVerifierTests.java
@@ -1235,6 +1235,42 @@ public class StepVerifierTests {
 	}
 
 	@Test
+	public void testWithDescriptionAndScenarioName() {
+		StepVerifierOptions options = StepVerifierOptions.create()
+		                                                 .initialRequest(3)
+		                                                 .scenarioName("some scenario name");
+		StepVerifier stepVerifier = StepVerifier
+				.create(Flux.just("foo", "bar", "baz"), options)
+				.expectNext("foo")
+				.as("first")
+				.expectNext("bar")
+				.as("second")
+				.expectNext("bar")
+				.as("third")
+				.as("this is ignored")
+				.expectComplete()
+				.log();
+
+		assertThatExceptionOfType(AssertionError.class)
+				.isThrownBy(stepVerifier::verify)
+				.withMessage("[some scenario name] expectation \"third\" failed (expected value: bar; actual value: baz)");
+	}
+
+	@Test
+	public void testDurationFailureWithScenarioName() {
+		StepVerifierOptions options = StepVerifierOptions.create()
+		                                                 .scenarioName("some scenario name");
+		StepVerifier stepVerifier = StepVerifier
+				.create(Mono.delay(Duration.ofMillis(100)), options)
+				.expectNextCount(1)
+				.expectComplete();
+
+		assertThatExceptionOfType(AssertionError.class)
+				.isThrownBy(() -> stepVerifier.verify(Duration.ofMillis(10)))
+				.withMessageStartingWith("[some scenario name] VerifySubscriber timed out on reactor.core.publisher.MonoDelay$MonoDelayRunnable@");
+	}
+
+	@Test
 	public void noCancelOnUnexpectedErrorSignal() {
 		LongAdder cancelled = new LongAdder();
 		assertThatExceptionOfType(AssertionError.class)

--- a/src/docs/asciidoc/testing.adoc
+++ b/src/docs/asciidoc/testing.adoc
@@ -127,6 +127,22 @@ TIP: By default, the `verify()` method and derived shortcut methods (`verifyThen
 `StepVerifier.setDefaultTimeout(Duration)` to globally set a timeout for these methods,
 or specify one on a per-call basis with `verify(Duration)`.
 
+=== Better identifying test failures
+`StepVerifier` provides two options to better identify exactly which expectation step caused
+a test to fail:
+
+ - `as(String)`: this method can be used **after** most `expect*` methods to give a description
+ to the preceding expectation. If the expectation fails, its error message will contain the
+ description. Terminal expectations and `verify` cannot be described that way.
+ - `StepVerifierOptions.create().scenarioName(String)`: Using `StepVerifierOptions` to create
+ you `StepVerifier`, you can use the `scenarioName` method to give the whole scenario a
+ name, which will also be used in assertion error messages.
+
+Note that in both cases, the use of the description/name in messages is only guaranteed for
+`StepVerifier` methods that produce their own `AssertionError` (eg. throwing an exception
+manually or through an assertion library in `assertNext` won't add the description/name to
+said error's message).
+
 == Manipulating Time
 
 `StepVerifier` can be used with time-based operators to avoid long run times for

